### PR TITLE
Create count_Inversions_merge_sort.cpp -- counts number of inversions using merge sort

### PR DESCRIPTION
**Pull request by rakshitraj**
Program to count the number of inversions in an array using Merge Sort. Inversions determine how close an array is to being sorted.

#### Checklist

- [x] Added description of change
- [x] Added file name matches [File name guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#New-File-Name-guidelines)

- [x] Added documentation so that the program is self-explanatory and educational - [Doxygen guidelines](https://www.doxygen.nl/manual/docblocks.html)
- [x] Relevant documentation/comments is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#Commit-Guidelines)
- [x] Search previous suggestions before making a new one, as yours may be a duplicate.
- [x] I acknowledge that all my contributions will be made under the project's license.

Notes: 

<a href="https://gitpod.io/#https://github.com/TheAlgorithms/C-Plus-Plus/pull/1078"></a>

<a href="https://gitpod.io/#https://github.com/TheAlgorithms/C-Plus-Plus/pull/1078"></a>

<a href="https://gitpod.io/#https://github.com/TheAlgorithms/C-Plus-Plus/pull/1078"></a>

<a href="https://gitpod.io/#https://github.com/TheAlgorithms/C-Plus-Plus/pull/1078"></a>

### CODE REVIEW

File: sorting/count_Inversions_merge_sort.cpp
Comment (Panquesito7) [inline sorting/count_Inversions_merge_sort.cpp:5]:
```suggestion
 * @details
 * Program to count the number of inversions in an array
```

File: sorting/count_Inversions_merge_sort.cpp
Comment (Panquesito7) [inline sorting/count_Inversions_merge_sort.cpp:54]: Missing return statements documentation.
See #962 as a reference. 🙂

File: sorting/count_Inversions_merge_sort.cpp
Comment (Panquesito7) [inline sorting/count_Inversions_merge_sort.cpp:74]: Same here as above.

File: sorting/count_Inversions_merge_sort.cpp
Comment (Panquesito7) [inline sorting/count_Inversions_merge_sort.cpp:107]: Same here as above.

File: sorting/count_Inversions_merge_sort.cpp
Comment (Panquesito7) [inline sorting/count_Inversions_merge_sort.cpp:142]: Don't use namespace-directives.
Use using-declarations instead.

File: sorting/count_Inversions_merge_sort.cpp
Comment (Panquesito7) [inline sorting/count_Inversions_merge_sort.cpp:150]: This shall be inside the `inversion` namespace.

File: sorting/count_Inversions_merge_sort.cpp
Comment (Panquesito7) [inline sorting/count_Inversions_merge_sort.cpp:159]:
```suggestion
/*
 * @brief Main function
 * @returns 0 on exit
 */
```

File: sorting/count_Inversions_merge_sort.cpp
Comment (Panquesito7) [inline sorting/count_Inversions_merge_sort.cpp:178]: You can add `@author` at the first comment block.

File: sorting/count_Inversions_merge_sort.cpp
Comment (Panquesito7) [inline sorting/count_Inversions_merge_sort.cpp:160]:
```suggestion
int main()
```

File: sorting/count_Inversions_merge_sort.cpp
Comment (Panquesito7) [inline sorting/count_Inversions_merge_sort.cpp:27]:
```suggestion
 *   5. Print the answer
 *
 * @author [Rakshit Raj](https://github.com/rakshitraj)
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]

File: sorting/count_Inversions_merge_sort.cpp
Comment (Panquesito7) [inline sorting/count_Inversions_merge_sort.cpp:56]:
```suggestion
 * @returns total number of inversions
```

File: sorting/count_Inversions_merge_sort.cpp
Comment (Panquesito7) [inline sorting/count_Inversions_merge_sort.cpp:77]:
```suggestion
 * @returns  total number of inversions to calling function
```

File: sorting/count_Inversions_merge_sort.cpp
Comment (Panquesito7) [inline sorting/count_Inversions_merge_sort.cpp:113]:
```suggestion
 * @returns number of inversions per merge step
```

File: sorting/count_Inversions_merge_sort.cpp
Comment (Panquesito7) [inline sorting/count_Inversions_merge_sort.cpp:152]:
```suggestion
 * @param arr array to print
```

File: sorting/count_Inversions_merge_sort.cpp
Comment (Panquesito7) [inline sorting/count_Inversions_merge_sort.cpp:153]:
```suggestion
 * @param array_size    size of input array
```

File: sorting/count_Inversions_merge_sort.cpp
Comment (Panquesito7) [inline sorting/count_Inversions_merge_sort.cpp:3]: Provide a Wikipedia link if possible.

File: sorting/count_Inversions_merge_sort.cpp
Comment (Panquesito7) [inline sorting/count_Inversions_merge_sort.cpp:37]:
```suggestion
/**
 * @namespace sorting
 * @brief Sorting algorithms
 */
namespace sorting {
/**
 * @namespace inversion
 * @brief Functions for Counting Inversions using Merge Sort
 */
namespace inversion {
```

File: sorting/count_Inversions_merge_sort.cpp
Comment (Panquesito7) [inline sorting/count_Inversions_merge_sort.cpp:164]:
```suggestion
}  // namespace inversion
}  // namespace sorting
```

File: sorting/count_Inversions_merge_sort.cpp
Comment (Panquesito7) [inline sorting/count_Inversions_merge_sort.cpp:179]:
```suggestion
    sorting::inversion::show(arr, size);
```

File: sorting/count_Inversions_merge_sort.cpp
Comment (Panquesito7) [inline sorting/count_Inversions_merge_sort.cpp:183]:
```suggestion
    std::cout << "\nThe number of inversions: "<< sorting::inversion::countInversion(arr) << "\n";
```

File: sorting/count_Inversions_merge_sort.cpp
Comment (Panquesito7) [inline sorting/count_Inversions_merge_sort.cpp:167]:
```suggestion
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]

**Review by rakshitraj — COMMENTED** [review]
Added suggestions from review

File: sorting/count_Inversions_merge_sort.cpp
Comment (Panquesito7) [inline sorting/count_Inversions_merge_sort.cpp:3]:
```suggestion
 * @brief Counting Inversions using [Merge Sort](https://en.wikipedia.org/wiki/Merge_Sort).
```

File: sorting/count_Inversions_merge_sort.cpp
Comment (Panquesito7) [inline sorting/count_Inversions_merge_sort.cpp:44]:
```suggestion
 * @brief Functions for Counting Inversions using [Merge Sort](https://en.wikipedia.org/wiki/Merge_Sort).
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
Please fix [`clang-tidy`](https://github.com/TheAlgorithms/C-Plus-Plus/pull/1078/checks?check_run_id=1109163123) warnings.

File: sorting/count_inversions.cpp
Comment (Panquesito7) [inline sorting/count_inversions.cpp:3]: Provide a Wikipedia link in Markdown format if available.

File: sorting/count_inversions.cpp
Comment (Panquesito7) [inline sorting/count_inversions.cpp:34]:
```suggestion
/**
 * @namespace sorting
 * @brief Sorting algorithms
 */
namespace sorting  {
/**
 * @namespace inversion
 * @brief Functions for Merge Sort algorithm
 */
namespace inversion {
```

File: sorting/count_inversions.cpp
Comment (Panquesito7) [inline sorting/count_inversions.cpp:156]: Please remove this and use namespace directly.

File: sorting/count_inversions.cpp
Comment (Panquesito7) [inline sorting/count_inversions.cpp:158]:
```suggestion
/**
 * @brief Main function
 * @returns 0 on exit
 */
int main() {
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]

**Comment by rakshitraj** [comment]
That was educational. I believe it is now ready to merge.

File: sorting/count_inversions.cpp
Comment (rakshitraj) [reply sorting/count_inversions.cpp:156]: Using namespace directly generates clang-tidy errors

Comment (Panquesito7) [reply sorting/count_inversions.cpp:156]: If you're going to call a function inside those namespaces, you need to add `sorting::inversion` before the function's name.

Comment (rakshitraj) [reply sorting/count_inversions.cpp:156]: So i need to remove the `using` directives and declare and call functions as <namespace>::<function_name>( args ) ?

File: sorting/count_inversions.cpp
Comment (Panquesito7) [inline sorting/count_inversions.cpp:105]: Missing return statement documentation.
```suggestion
 * @param right  upper bound of arr[] and right-sub-array
 * @returns description
```

File: sorting/count_inversions.cpp
Comment (Panquesito7) [inline sorting/count_inversions.cpp:73]: Missing return statement documentation.
```suggestion
 * @param right - upper bound of array
 * @returns description
```

File: sorting/count_inversions.cpp
Comment (Panquesito7) [inline sorting/count_inversions.cpp:141]:
```suggestion
 * @param array_size    size of input array arr[]
 * @returns void
```

File: sorting/count_inversions.cpp
Comment (Panquesito7) [reply sorting/count_inversions.cpp:156]: When registering a function inside a namespace, you don't need to add the `sorting::inversion` at the beginning.
When calling the function itself, it **is** needed to add that at the beginning. Otherwise, it won't work.

Example:

```c++
// Registering namespace
namespace sorting {
void function(int arg_1, int arg_2) {/** definition */ }
} // namespace sorting

int main() {
    int arg_1, arg_2 = 0;
    sorting::function(arg_1, arg_2); // Calling the function inside a namespace
}
```

File: sorting/count_inversions.cpp
Comment (Panquesito7) [inline sorting/count_inversions.cpp:36]: Not needed to add the namespace names at the beginning if these functions are inside the namespace.

File: sorting/count_inversions.cpp
Comment (rakshitraj) [reply sorting/count_inversions.cpp]: Noted. Fixing all in the next commit.

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
https://github.com/TheAlgorithms/C-Plus-Plus/pull/1078#discussion_r497854887

File: sorting/count_inversions.cpp
Comment (Panquesito7) [inline sorting/count_inversions.cpp:205]: This shall be added in the first comment block.
```suggestion
```

File: sorting/count_inversions.cpp
Comment (Panquesito7) [inline sorting/count_inversions.cpp:28]:
```suggestion
 *
 * @author [Rakshit Raj](https://github.com/rakshitraj)
```

File: sorting/count_inversions.cpp
Comment (Panquesito7) [inline sorting/count_inversions.cpp:164]: Please add self-test cases.
See #962 as a reference. 🙂

File: sorting/count_inversions.cpp
Comment (Panquesito7) [inline sorting/count_inversions.cpp:5]:
```suggestion
 * @details
 * Program to count the number of inversions in an array
```

File: sorting/count_inversions.cpp
Comment (Panquesito7) [inline sorting/count_inversions.cpp:15]:
```suggestion
 * ### Algorithm
 
```

**Review by Panquesito7 — COMMENTED** [review]
Please enable [GitHub Actions](https://docs.github.com/en/free-pro-team@latest/actions/learn-github-actions) in your repository of this fork in this link: https://github.com/rakshitraj/C-Plus-Plus/actions
This will enable enhanced formatting and code. To trigger the GitHub Actions, push an small commit. For example, remove line 44. Next, we'll see a commit by GitHub Actions updating `DIRECTORY.md`. 🙂

File: sorting/count_inversions.cpp
Comment (rakshitraj) [reply sorting/count_inversions.cpp:164]: added 3 test cases

File: sorting/count_inversions.cpp
Comment (Panquesito7) [inline sorting/count_inversions.cpp:237]: This shall be added in the first comment block, as I said above.

File: sorting/count_inversions.cpp
Comment (Panquesito7) [inline sorting/count_inversions.cpp:166]:
```suggestion
 * @brief Test implementations
```

File: sorting/count_inversions.cpp
Comment (Panquesito7) [inline sorting/count_inversions.cpp:15]:
```suggestion
 * ### Algorithm

```

File: sorting/count_inversions.cpp
Comment (Panquesito7) [inline sorting/count_inversions.cpp:237]:
```suggestion
```

File: sorting/count_inversions.cpp
Comment (Panquesito7) [inline sorting/count_inversions.cpp:234]:
```suggestion
    // Run test implementations
```

**Review by Panquesito7 — COMMENTED** [review]
Awesome work on updating the codes, looks amazing now. 😃 👍
IMHO, lines 193-227 should be removed. Seems to be the old code for `main()`.

File: sorting/count_inversions.cpp
Comment (rakshitraj) [reply sorting/count_inversions.cpp:237]: If its all the same, I'd like to keep it as reference for any beginner. But, I give you the final say. Say the word and I'll remove it ASAP.

Comment (rakshitraj) [reply sorting/count_inversions.cpp:253]: It refers to the void body() function which i have also commented out

Comment (Panquesito7) [reply sorting/count_inversions.cpp:253]: Now thinking on it, you can keep it if you wish. 🙂

**Review by Panquesito7 — APPROVED** [review]
LGTM, excellent work. 👍 😄

File: sorting/count_inversions.cpp
Comment (rakshitraj) [reply sorting/count_inversions.cpp:253]: I'll keep it. Thanks

**Comment by rakshitraj** [comment]
Thanks @Panquesito7 . I wish my college project mentor was as succint, to-the-point and knowledgeable as you. It has a pleasure working with you. I learned a lot!

**Comment by rakshitraj** [comment]
https://github.com/TheAlgorithms/C/blob/master/sorting/merge_sort.c will **only** sort the array and not count inversions. In order to count inversions changes will need to be made to both the MergeSort() and merge() steps. After each merge step and the number of inversions has to counted by the function internally.

**Comment by kvedala** [comment]
> https://github.com/TheAlgorithms/C/blob/master/sorting/merge_sort.c will **only** sort the array and not count inversions. In order to count inversions changes will need to be made to both the MergeSort() and merge() steps. After each merge step and the number of inversions has to counted by the function internally.
> 
> This addition functionality is bundled together in `sorting::inversion::countInversion()` function.

Agreed that the code in this PR is different. The code links are only for your *reference* to see if you can update the code to more C++ style.

**Comment by rakshitraj** [comment]
@kvedala Are there any changes that you would like to suggest?

**Comment by kvedala** [comment]
> @kvedala Are there any changes that you would like to suggest?

Yes, please update the code to use templates so that the code can be executed for signed unsigned integers, floats, doubles, derived classes, etc. Hence, the references like [this](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/sorting/merge_insertion_sort.cpp)

**Pull request closed without merging**
